### PR TITLE
Add Double and Int Convenience Properties

### PR DESCRIPTION
**Pull request by vermont42**
Building on the [work](https://github.com/apple/swift-syntax/pull/148/files) of @ahoppen on [SR-11580](https://bugs.swift.org/browse/SR-11580), this PR adds convenience properties to get `Double` and `Int` values from `FloatLiteralExprSyntax` and `IntegerLiteralExprSyntax`, respectively.

The two new properties, `floatingValue` and `intValue`, use failable initializers of `Double` and `Int`: `init?<S>(_ text: S) where S : StringProtocol` and `init?<S>(_ text: S, radix: Int) where S : StringProtocol`. These initializers fail if the `String`s passed in do not represent valid `Double`s or `Int`s. The question arises: what should `floatingValue` and `intValue` return, if anything, if initialization fails? I considered the following possibilities, ultimately choosing the fourth.

1. `floatingValue` could return `Double.nan`. This would make sense because, for example, ``"🍕.🌍"`` is truly not a number, floating-point or otherwise. The problem with this approach is that there is no corresponding `Int.nan`, which would prevent `floatingValue` and `intValue` from returning the same sort of value given invalid input.  `Decimal` _does_ have a [`nan`](https://developer.apple.com/documentation/foundation/decimal/1780019-nan) value, so if `intValue` returned a `Decimal` rather than an `Int`, that would solve the problem, but I, as a consumer of the API, would find this counter-intuitive because `"42"` in a source file represents an `Int`, not a `Decimal`.

2. The return types of `floatingValue` and `intValue` could be optional. If `Double` or `Int` initialization fails, `floatingValue` and `intValue` could therefore return `nil`. I preferred not to impose on API consumers the burden of unwrapping.

3. The return types of `floatingValue` and `intValue` could be `Result<Double, LiteralError>` and `Result<Int, LiteralError>`, respectively. If `Double` or `Int` initialization fails, `floatingValue` and `intValue` could return `.failure(.invalid)`. I preferred not to impose on API consumers the burden of `switch`ing on the returned value.

4. The result of the `Double` or `Int` initialization can be force-unwrapped. This causes a crash if initialization fails. Crashing is [considered harmful](https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf), though [not](https://t.co/30VaX2WWHg?amp=1) by everyone. Force-unwrapping is the most convenient option for API consumers, which is why I chose it. I observe that force-unwrapping does appear elsewhere in SwiftSyntax, for example [here](https://github.com/apple/swift-syntax/blob/27bc2a11a7723dfe9bc5bda3caee49d5f5e2df59/Sources/SwiftSyntax/Diagnostic.swift#L173), [here](https://github.com/apple/swift-syntax/blob/27bc2a11a7723dfe9bc5bda3caee49d5f5e2df59/Sources/SwiftSyntax/SyntaxOtherNodes.swift#L97), [here](https://github.com/apple/swift-syntax/blob/27bc2a11a7723dfe9bc5bda3caee49d5f5e2df59/Sources/SwiftSyntax/SyntaxData.swift#L308), and [here](https://github.com/apple/swift-syntax/blob/27bc2a11a7723dfe9bc5bda3caee49d5f5e2df59/Sources/SwiftSyntax/Syntax.swift#L326), and I suspect that the authors of this code made similar judgments about the convenience of force-unwrapping versus the crashing risk it imposes.

### CODE REVIEW

File: Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import SwiftSyntax
+
+public class SyntaxConvenienceMethodsTests: XCTestCase {
+
+  public func testFloatingValues() {
+    testFloatingValue(text: "5.3_8", expectedValue: 5.38)
+    testFloatingValue(text: "12e3", expectedValue: 12000.0)
+    testFloatingValue(text: "32E1", expectedValue: 320.0)
+    testFloatingValue(text: "0xdEFACE.C0FFEEp+1", expectedValue: 29226397.507810354)
+    testFloatingValue(text: "0xaffab1e.e1fP-2", expectedValue: 46131911.72064209)
```
Comment (vermont42) [inline Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift:11]: I could have used the following approach to avoid repetition of the `testFloatingValue()` call:

```
[
  ("5.3_8", 5.38),
  ("12e3", 12000.0),
  ("32E1", 320.0),
  ("0xdEFACE.C0FFEEp+1", 29226397.507810354),
  ("0xaffab1e.e1fP-2", 46131911.72064209)
].forEach {
  testFloatingValue(text: $0.0, expectedValue: $0.1)
}
```
Although this approach would have been [DRY](https://dzone.com/articles/software-design-principles-dry-and-kiss)er, I find that it makes failure diagnosis more difficult, so I did not use it.

**Comment by akyrtzi** [comment]
@vermont42 thanks for working on this! I would be fine with #4 (force-unwrap) since it can be considered a violation of tree invariants if an `IntegerLiteralExprSyntax` node doesn't actually contain an integer.

However, if we do force-unwrap we need to protect against the tree getting into such state in the first place. I think we can safely assume that the parser will not provide such a node (and if it does that's a parser bug), but another source of these are the APIs for creating syntax nodes programmatically. It is important to catch invariant breakage at the earliest source, so we should prevent a case of manually creating a `IntegerLiteralExprSyntax` node with a string that is not an integer.

That may require changing the python code on the swift repo side that does the code-generation for the programmatic APIs. e.g. maybe add a flag to indicate that the programmatic API for a node will be manually written on the SwiftSyntax side and avoid code-generating, then we write the functions manually and include the validation aspect.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+public extension IntegerLiteralExprSyntax {
+  var integerValue: Int {
+    let text = self.digits.text
+    let type = IntegerType(text: text)
+    let textWithoutPrefix: String
+    let digitsStartIndex = text.index(text.startIndex, offsetBy: type.prefixLength())
+    textWithoutPrefix = String(text.suffix(from: digitsStartIndex))
+
+    let textWithoutPrefixOrUnderscores = textWithoutPrefix.filter {
+      $0 != "_"
```
Comment (rintaro) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:31]: Technically, `_` in each component is not valid. e.g.  `0x_12`, `0._12` and `12e_3`. But I think it's OK to accept them for now.
@akyrtzi What do you think?

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    return Int(textWithoutPrefixOrUnderscores, radix: type.radix())!
+  }
+
+  private enum IntegerType {
```
Comment (rintaro) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:37]: I feel like this type is overkill. It could be a single inner function returning a tuple of `prefixLength` and `radix`

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 2 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:31]: I think it's fine but adding a FIXME comment here would be good.

File: Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  private func testFloatingValue(text: String, expectedValue: Double) {
+    let digits = SyntaxFactory.makeFloatingLiteral(text)
+    let literalExpr = SyntaxFactory.makeFloatLiteralExpr(floatingDigits: digits)
+    XCTAssertEqual(literalExpr.floatingValue, expectedValue)
+  }
```
Comment (rintaro) [inline Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift:18]: Just a stylistic request, could you move this `testFloatingValue` and `testIntegerValue` out of `SyntaxConvenienceMethodsTests` type? They are not "test" methods for XCTest and can be a free functions.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 2 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:31]: @rintaro I observe that `_12` and `0x_2` don't parse as `IntegerLiteralExprSyntax` or `FloatLiteralExprSyntax`, respectively. My properties therefore shouldn't get those sorts of input. Should the properties nonetheless `fatalError()` if the `_` is the first character after the base prefix (or is the first character in the case of base 10)?

Comment (rintaro) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:31]: As @akyrtzi said in another comment, we should prevent a case of creating syntax node with invalid literal text. If you do that, you don't need to check `_` in this `integerValue` or `floatingValue` accessor.

Also, when you check them, you need to check not only for the integral part, but you also need to check the fractional part and the exponent part. That's not so simple to implement in this PR. So I suggest you to add a `FIXME` comment as @akyrtzi said.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 3 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift]: Done.

File: Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift (hunk 4 above)
Comment (vermont42) [reply Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift]: Done.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public extension FloatLiteralExprSyntax {
+  var floatingValue: Double {
+    let floatingDigitsWithoutUnderscores = floatingDigits.text.filter {
+      $0 != "_"
+    }
+    return Double(floatingDigitsWithoutUnderscores)!
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:18]: I wouldn’t have thought that the `Double` initialiser can handle hex float literals. 

I think it’s odd though, that the `Double` initialiser is able to handle hex literals while the `Int` initialiser is not. It’s not really part of this PR but it might be worth filing a bug report against the standard library to either support hex literals in both or neither of them.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  }
+}
+
+public extension IntegerLiteralExprSyntax {
+  var integerValue: Int {
+    let text = self.digits.text
+    let (prefixLength, radix) = prefixLengthAndRadix(text: text)
+    let digitsStartIndex = text.index(text.startIndex, offsetBy: prefixLength)
+    let textWithoutPrefix = String(text.suffix(from: digitsStartIndex))
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:27]: I don’t think you need the `String` initialiser wrapping `text.suffix`. It should be more efficient if the `filter` in the next line directly works on the `Substring` returned by `suffix` instead of copying the substring data to a new `String` buffer.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:30]: I would say that for now it’s sufficient if we uphold the invariant that the contents will of a integer / float literal are always parsable from the `integerValue` resp. `floatingValue` properties, especially since the implementation is nice and short so far!

That should already be an improvement over the current behaviour and if we add a FIXME, we can come back to it and improve it later.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    let hexadecimalPrefix = "0x"
+
+    let binaryRadix = 2
+    let octalRadix = 8
+    let decimalRadix = 10
+    let hexadecimalRadix = 16
+
+    switch String(text.prefix(nonDecimalPrefixLength)) {
+    case binaryPrefix:
+      return (nonDecimalPrefixLength, binaryRadix)
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:51]: What about using `binaryPrefix.count` instead of the `nonDecimalPrefixLength` variable? Same below.

File: Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift (hunk 1 above)
Comment (ahoppen) [reply Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift:11]: The testing approach seems good to me. One idea though (I’m not quite sure if it’s an improvement or not) would be to use the literal that you’re specifying in the String also in the Swift source code, i.e.\

```swift
testFloatingValue(text: "0xaffab1e.e1fP-2", expectedValue: 0xaffab1e.e1fP-2)
```

Mostly because my mental hex parsing skills aren't brilliant and I’m not quite sure if `0xaffab1e.e1fP-2` is really equal to `46131911.72064209`.

**Comment by ahoppen** [comment]
> However, if we do force-unwrap we need to protect against the tree getting into such state in the first place. I think we can safely assume that the parser will not provide such a node (and if it does that's a parser bug), but another source of these are the APIs for creating syntax nodes programmatically. It is important to catch invariant breakage at the earliest source, so we should prevent a case of manually creating a IntegerLiteralExprSyntax node with a string that is not an integer.

@akyrtzi I’ve just been digging around the generated code to give @vermont42 some advice on what needs to be modified. While doing so I have been thinking what the desired API would be. I would argue that making all methods that construct a integer/float literal return an `Optional` would be the cleanest solution, even though it means that these are the only types with a failable initialiser. What do you think?

**Comment by akyrtzi** [comment]
>  I would argue that making all methods that construct a integer/float literal return an Optional would be the cleanest solution, even though it means that these are the only types with a failable initialiser. What do you think?

That makes sense to me 👍

File: Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift (hunk 1 above)
Comment (vermont42) [reply Tests/SwiftSyntaxTest/SyntaxConvenienceMethodsTests.swift]: Done. As an aside, [this](https://nedbatchelder.com/text/hexwords.html) is my go-to source for fun hex words.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 7 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:51]: Done.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 6 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift]: Done.

**Comment by vermont42** [comment]
### Update

I have partially implemented the suggestion of upholding the invariant. Specifically, I modified the `public init?(_ syntax: Syntax) {` initializer to validate but am still figuring out how to modify the `internal init(_ data: SyntaxData) {` initializer.

The invariant code relies on the following non-breaking code changes in the main Swift repo. Do reviewers agree that, after this SwiftSyntax PR is good to go, I should PR the main Swift repo with the Python changes, wait for that to land, and then request that the SwiftSyntax PR be merged?

`ExprNodes.py`:

```
    Node('FloatLiteralExpr', kind='Expr',
         children=[
             Child('FloatingDigits', kind='FloatingLiteralToken'),
         ],
         must_uphold_invariant=True),     // NEW
```
```
    Node('IntegerLiteralExpr', kind='Expr',
         children=[
             Child('Digits', kind='IntegerLiteralToken'),
         ],
         must_uphold_invariant=True),     // NEW
```
`Node.py`:
```
class Node(object):
   [omitted]
    def __init__(self, name, description=None, kind=None, traits=None,
                 children=None, element=None, element_name=None,
                 element_choices=None, omit_when_empty=False,
                 must_uphold_invariant=False):     // NEW
       [omitted]
        self.must_uphold_invariant = must_uphold_invariant     // NEW
```

**Comment by vermont42** [comment]
### Update

I pushed a commit causing the `internal init(_ data: SyntaxData) {` initializer to perform validation and be failable.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+public extension ExprSyntaxProtocol {
+  var isValidLiteral: Bool {
+    if let integerLiteralExprSyntax = self as? IntegerLiteralExprSyntax {
+      return integerLiteralExprSyntax.potentialIntegerValue != nil
+    } else if let floatLiteralExprSyntax = self as? FloatLiteralExprSyntax {
+      return floatLiteralExprSyntax.potentialFloatingValue != nil
+    } else {
+      return false
+    }
+  }
```
Comment (rintaro) [inline Sources/SwiftSyntax/SyntaxConvenienceMethods.swift:79]: `isValidLiteral` is too specific to these types. Could you make it more general like `isValid`, `validate()` or something?
Also, I think you should split this function for each types like:
```swift
extension IntegerLiteralExprSyntax {
  func validate() -> Bool { potentialIntegerValue != nil }
}
extension FloatLiteralExprSyntax {
  func _validate() -> Bool { potentialFloatingValue != nil }
}
```
This way, compiler can check that the validation function is actually implemented for every `must_uphold_invariant` node.

File: Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift [hunk 9]
```diff
@@ -626,15 +626,15 @@ public enum SyntaxFactory {
...
   }
 
   public static func makeBlankFloatLiteralExpr() -> FloatLiteralExprSyntax {
     let data = SyntaxData.forRoot(RawSyntax.create(kind: .floatLiteralExpr,
       layout: [
       RawSyntax.missingToken(TokenKind.floatingLiteral("")),
     ], length: .zero, presence: .present))
-    return FloatLiteralExprSyntax(data)
+    return FloatLiteralExprSyntax(data)!
   }
```
Comment (rintaro) [inline Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift:638]: I believe this always crashes because `.floatingLiteral` token is constructed with an empty string.
IMO this function is completely useless in the first place. Could you avoid auto generating this and `makeBlankIntegerLiteralExpr`?

File: Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift [hunk 10]
```diff
@@ -626,15 +626,15 @@ public enum SyntaxFactory {
     let raw = RawSyntax.createAndCalcLength(kind: SyntaxKind.floatLiteralExpr,
       layout: layout, presence: SourcePresence.present)
     let data = SyntaxData.forRoot(raw)
-    return FloatLiteralExprSyntax(data)
+    return FloatLiteralExprSyntax(data)!
```
Comment (rintaro) [inline Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift:629]: This crashes if the argument `floatingDigits` is invalid.
Could you make this `makeFloatLiteralExpr` and `makeIntegerLiteralExpr` failable (i.e. return `Optional`)?

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift (hunk 8 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift]: Done. Following the `isEmpty` example in the API Design Guidelines, I chose to leave it as a computed property and spell it `isValid`.

File: Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift (hunk 9 above)
Comment (vermont42) [reply Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift]: Done.

File: Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift (hunk 10 above)
Comment (vermont42) [reply Sources/SwiftSyntax/gyb_generated/SyntaxFactory.swift]: Done. Getting the Python indentation in `SyntaxFactory.swift.gyb` right was tricky. 😅

File: Sources/SwiftSyntax/SyntaxFactory.swift.gyb [hunk 11]
```diff
@@ -82,7 +86,7 @@ public enum SyntaxFactory {
   }
 %   end
 
-%   if not node.is_base():
+%   if not node.is_base() and not node.must_uphold_invariant:
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxFactory.swift.gyb:89]: Technically, I don’t think it’s correct to remove the `makeBlank` method if a node must uphold an invariant. 

In the future we might add invariants to other nodes as well that could have a valid blank content. One example that jumps to my mind would be string literals. For consistency, I would rather have this method return an Optional and return `nil` if the node kind has no valid blank representation.

@akyrtzi What do you think?

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template [hunk 12]
```diff
@@ -70,14 +70,28 @@ public struct ${node.name}: ${base_type}Protocol, SyntaxHashable {
   public init?(_ syntax: Syntax) {
     guard syntax.raw.kind == .${node.swift_syntax_kind} else { return nil }
     self._syntaxNode = syntax
+%   if node.must_uphold_invariant:
+    if !isValid {
+      return nil
+    }
+%   end
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:77]: I think it’s save to make it a `fatalError` if the node is not valid at this stage. By our invariant, we know that every `Syntax` node is valid and casting it should not change the invariant.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template [hunk 13]
```diff
@@ -70,14 +70,28 @@ public struct ${node.name}: ${base_type}Protocol, SyntaxHashable {
...
   }
 
   /// Creates a `${node.name}` node from the given `SyntaxData`. This assumes
   /// that the `SyntaxData` is of the correct kind. If it is not, the behaviour
   /// is undefined.
+%   if node.must_uphold_invariant:
+  internal init?(_ data: SyntaxData) {
+%   else:
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:85]: Could you also update the comment to say that the initialiser returns `nil` if the invariant is not satisfied?

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template [hunk 14]
```diff
@@ -156,7 +170,11 @@ public struct ${node.name}: ${base_type}Protocol, SyntaxHashable {
     let raw = newChild?.raw ?? ${make_missing_swift_child(child)}
 %       end
     let newData = data.replacingChild(raw, at: Cursor.${child.swift_name})
+%       if node.must_uphold_invariant:
+    return ${node.name}(newData)!
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:174]: We definitely shouldn‘t force unwrap here. The invariant will not be satisfied if the user passes in a child that does not satisfy the invariant (e.g. an invalid integer literal). Instead, this method should return an Optional.

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb [hunk 15]
```diff
@@ -91,7 +91,11 @@ open class SyntaxRewriter {
       if let newNode = visitAny(node._syntaxNode) { return newNode }
       return Syntax(visit(node))
 %   else:
+%     if node.must_uphold_invariant:
+      let node = ${node.name}(data)!
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxRewriter.swift.gyb:95]: Could you add a comment like the following here to document why the use of the `!` is safe here?

```swift
// We know that the SyntaxData is valid since we are walking a valid syntax tree and haven't modified the syntax data.  Thus the below initialiser will never return nil.
```

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb [hunk 16]
```diff
@@ -86,7 +86,11 @@ open class SyntaxVisitor {
       }
       visitPost(node)
 %   else:
+%     if node.must_uphold_invariant:
+      let node = ${node.name}(data)!
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:90]: Could you also add a comment like the one I suggested for `SyntaxRewriter` here?

**Review by ahoppen — COMMENTED** [review]
Overall looking really good. Congrats on getting the entire gyb-generation pipeline to work.

I’ve posted a couple comments on the handling of optionals inline but they should be easy to fix.

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb (hunk 15 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxRewriter.swift.gyb:95]: Done.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template (hunk 13 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:86]: Done.

File: Sources/SwiftSyntax/SyntaxVisitor.swift.gyb (hunk 16 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxVisitor.swift.gyb:91]: Done.

**Comment by vermont42** [comment]
I implemented @ahoppen's comment suggestions but will hold off on implementing the other suggestions in order to give @rintaro and @akyrtzi the opportunity to respond.

File: Sources/SwiftSyntax/SyntaxFactory.swift.gyb (hunk 11 above)
Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxFactory.swift.gyb:89]: Just to step back a bit here, what is the use case for the `makeBlank` methods, why do we need them? When would I call a `makeBlank` method instead of explicitly creating the syntax node that I want?

Comment (akyrtzi) [reply Sources/SwiftSyntax/SyntaxFactory.swift.gyb:89]: Paging @harlanhaskins for my question.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template (hunk 12 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:77]: Done.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template (hunk 14 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template]: Done.

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb (hunk 15 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxRewriter.swift.gyb:95]: I think you forgot to add the comment ;-)

**Review by ahoppen — COMMENTED** [review]
Thanks for yet another round and your persistence ;-)

I have looked through the code again. I’ve got two more nitpicks and just realised that setting an invalid text as the contents of an integer literal crashes. I’ve added some comments inline. 

Other than that, I think we’re really close.

File: Sources/SwiftSyntax/SyntaxFactory.swift.gyb (hunk 11 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SyntaxFactory.swift.gyb:89]: Ping @harlanhaskins

File: Sources/SwiftSyntax/SyntaxRewriter.swift.gyb (hunk 15 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxRewriter.swift.gyb:96]: Sorry about that. Fixed.

File: Sources/SwiftSyntax/SyntaxConvenienceMethods.swift [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    let textWithoutPrefix = text.suffix(from: digitsStartIndex)
+
+    let textWithoutPrefixOrUnderscores = textWithoutPrefix.filter {
+      $0 != "_"
+    }
+
+    return Int(textWithoutPrefixOrUnderscores, radix: radix)
+  }
+
+  static func prefixLengthAndRadix(text: String) -> (Int, Int) {
```
[the remark this answers is not in the document]
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxConvenienceMethods.swift]: Done. (For real. 😅)

File: Sources/SwiftSyntax/SyntaxFactory.swift.gyb (hunk 11 above)
Comment (harlanhaskins) [reply Sources/SwiftSyntax/SyntaxFactory.swift.gyb:89]: Oh sorry! I think `makeBlank` was meant to be used for e.g. error recovery where we'd just fill in missing tokens with blank values in order to send a well-formed-but-invalid AST to semantic analysis.

**Comment by vermont42** [comment]
@ahoppen I removed the two setters and converted them into throwing functions, preferring that approach over converting all setters to throwing functions, notwithstanding the inconsistency, because throwing functions are less pleasant for callers to use than garden-variety setters and are therefore to be avoided when possible. I think some inconsistency between, on the one hand,`FloatLiteralExprSyntax` and `IntegerLiteralExprSyntax`, and, on the other, everything else is actually warranted because of the constraints on validity to which these two types, but not others, are subject. I also added the test you mentioned and updated the changelog.

I am unclear on what, if anything, still needs to be done about `Blank`s. I can either leave the current approach of not generating them or restore them and return `nil`.

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template [hunk 18]
```diff
@@ -122,14 +122,33 @@ public struct ${node.name}: ${base_type}Protocol, SyntaxHashable {
...
+    public var description: String {
+      switch self {
+      case .invalid(let ${child.swift_name}):
+        return "attempted to use setter with invalid ${child.name} \"\(${child.swift_name})\""
+      }
+    }
+  }
+
+  mutating public func set${child.name}(_ ${child.swift_name}: ${ret_type}) throws {
+    if let tokenSyntax = with${child.name}(${child.swift_name}) {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template:145]: I think technically this variable shouldn’t be called `tokenSyntax` because the child doesn’t need to be a token (although it is in both of our cases). I would prefer something like `childSyntax` or something along those lines.

File: Tests/SwiftSyntaxTest/SyntaxTests.swift [hunk 19]
```diff
@@ -144,4 +144,57 @@ public class SyntaxTests: XCTestCase {
...
+    try! integerExpr.setDigits(token)
+    XCTAssert(integerExpr.isValid)
+
+    let invalidNumericLiteral = "🥥"
+
+    source = invalidNumericLiteral
+    tree = try! SyntaxParser.parse(source: source)
+    token = tree.firstToken!
+    do {
+      try integerExpr.setDigits(token)
```
Comment (ahoppen) [inline Tests/SwiftSyntaxTest/SyntaxTests.swift:167]: I think there’s `XCTAssertThrowsError` to test that an error is thrown. This test would also pass if `setDigits` doesn’t throw.

File: Tests/SwiftSyntaxTest/SyntaxTests.swift [hunk 20]
```diff
@@ -144,4 +144,57 @@ public class SyntaxTests: XCTestCase {
...
+    tree = try! SyntaxParser.parse(source: source)
+    token = tree.firstToken!
+    try! floatExpr.setFloatingDigits(token)
+    XCTAssert(floatExpr.isValid)
+
+    source = invalidNumericLiteral
+    tree = try! SyntaxParser.parse(source: source)
+    token = tree.firstToken!
+    do {
+      try floatExpr.setFloatingDigits(token)
```
Comment (ahoppen) [inline Tests/SwiftSyntaxTest/SyntaxTests.swift:191]: Same here. Use `XCTAssertThrowsError`.

**Review by ahoppen — COMMENTED** [review]
Nice one. I’ve got two minor comments inline that should be super easy to fix.

Regarding the `Blank` methods: I’m fine with the way it is right now. We can discuss what to do with them in a follow-up PR to get this one landed.

Could you also open a PR in the main Swift repos with your changes to `swift_syntax_support`? There’s no need to write a long description there, just reference this PR. Afterwards I can throw CI against your changes and we will see what she has to say.

Also: Before merging, could you squash your changes into one or two (or whatever seems reasonable to you) commits?

File: Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template (hunk 18 above)
Comment (vermont42) [reply Sources/SwiftSyntax/SyntaxNodes.swift.gyb.template]: Done.

File: Tests/SwiftSyntaxTest/SyntaxTests.swift (hunk 19 above)
Comment (vermont42) [reply Tests/SwiftSyntaxTest/SyntaxTests.swift]: Done.

File: Tests/SwiftSyntaxTest/SyntaxTests.swift (hunk 20 above)
Comment (vermont42) [reply Tests/SwiftSyntaxTest/SyntaxTests.swift]: Done.

**Comment by vermont42** [comment]
@ahoppen I `squash`ed all but my first commit, which I `reword`ed to `Implements SR-11580`. I'm not sure why Doug's commit appears in my commit history because I left that as `pick`.
Screen Shot 2020-09-14 at 3 04 00 PM

There were some conflicts with the the main repo's `master` branch, so I resolved those.

Please let me know if further `git` actions should be taken. 😅

**Comment by ahoppen** [comment]
> @ahoppen I squashed all but my first commit, which I reworded to Implements SR-11580. I'm not sure why Doug's commit appears in my commit history because I left that as pick.

I think Doug’s commit should disappear if you rebase your PR on the current master branch. That way you should also be able to remove the merge commit.

**Comment by vermont42** [comment]
At @ahoppen's request, I `squash`ed my changes. Due to an off-by-one error in my interactive `rebase`, the extraneous commit shown below appeared in the PR's commit history. After a couple of hours tinkering with `git` commands, I realized that the cleanup task is beyond my ability to implement, as demonstrated by the current, polluted state of this PR's commit history. I am therefore closing this PR and raising a new one with a `squash`ed, unpolluted commit history. I apologize for the obscuring of feedback on the about-to-be-closed PR that this action will cause. The new PR will reference this PR.

Screen Shot 2020-09-17 at 9 12 42 PM
